appsrc/acrnema/andump.man
.TH ANDUMP 1 "6 March 2014" "DICOM PS3" "DICOM PS3 - Describe ACR-NEMA file content"
.SH NAME
andump \- ACR/NEMA DICOM PS3 ... Describe ACR-NEMA file content
.SH SYNOPSIS
.HP 10
.B andump
.so man1/gen.so
[
.B \-ancreate
]
[
.B \-extractCSA
]
[
.B \-showoffset|-showoffset-hex
|-showoffset-octal|-showoffset-oct
|-showoffset-decimal|-showoffset-dec
]
[
.B \-filename
]
.so man1/optin.so
.SH DESCRIPTION
.LP
.B andump
reads the named acr-nema input file and describes the information contained, without attempting to interpret the structure of the message (cf. dcdump).
.LP
The group and element number, dictionary and explicit value representation, description of tag, value length and value of the element are displayed, 
optionally with an offset byte count from the start of the file.
.SH OPTIONS
The attribute values, description and verbose output go to standard error.
.LP
Binary attributes are written in hexadecimal with a preceding
"0x". Numeric string attributes are written in decimal. Attribute values
are displayed in hexadecimal or string format as determined by the value representation.
.PP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-ancreate
.RS
Display item, end item and end sequence delimiter tags in a form suitable to feed into ancreate.
.RE
.TP
.B \-extractCSA
.RS
Extract the DICOM-like attributes buried within the Siemens private CSA header OB value.
.RE
.TP
.B \-showoffset|-showoffset-hex
.RS
Display byte offset from file start in hexadecimal.
.RE
.TP
.B \-showoffset-octal|-showoffset-oct
.RS
Display byte offset from file start in octal.
.RE
.TP
.B \-showoffset-decimal|-showoffset-dec
.RS
Display byte offset from file start in decimal.
.RE
.TP
.B \-filename
.RS
Show the name of the file supplied in the arguments; a hyphen will be reported if no filename was supplied.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
% andump NM.dc3
.RE
(0x0008,0x0000) UL Group Length  VR=<UL> VL=<0x0004> [0x08] 
.RE
(0x0008,0x0008) CS Image Type    VR=<CS> VL=<0x0024> <...> 
.RE
...
.RE
(0x0054,0x0013) SQ ... Sequence  VR=<SQ> VL=<0xffffffff> []
.RE
(0xfffe,0xe000) NONE Item        VR=<> VL=<0xffffffff> []
.RE
(0x0054,0x0014) DS Energy Win LL VR=<DS> VL=<0x0006> <1.0 > 
.RE
(0x0054,0x0015) DS Energy Win UL VR=<DS> VL=<0x0006> <9.0 > 
.RE
(0xfffe,0xe00d) NONE Item Del .. VR=<> VL=<0x0000> []
.RE
(0xfffe,0xe0dd) NONE Seq Del ..  VR=<> VL=<0x0000> []
.RE
...
.RE
\ 
.RE
\ 
.RE
% andump -ancreate NM.dc3
.RE
(0x0008,0x0000) UL Group Length  VR=<UL> VL=<0x0004> [0x08] 
.RE
(0x0008,0x0008) CS Image Type    VR=<CS> VL=<0x0024> <...> 
.RE
...
.RE
(0x0054,0x0013) SQ ... Sequence  VR=<SQ> VL=<0xffffffff> []
.RE
%item
.RE
(0x0054,0x0014) DS Energy Win LL VR=<DS> VL=<0x0006> <1.0 > 
.RE
(0x0054,0x0015) DS Energy Win UL VR=<DS> VL=<0x0006> <9.0 > 
.RE
%enditem
.RE
%endseq
.RE
...
.RE
\ 
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR ancreate(1) ,
.BR dcdump(1) ,
.BR dcintro(1)
.SH AUTHOR
Copyright (c) 1993-2024, David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
241220:	libsrc/src/dctool/attrtypv.cc: Allow TAB character in control characters for text VRs (CP 1425)

241114:	appsrc/dcfile/dciodvfy.cc: Allow single byte of padding for odd number of 8 bit LUT entries in 16 bit words (000607)

241114:	appsrc/dcfile/dciodvfy.cc: Remove check on SegmentNumber monotonically increasing from 1 (not 0) if SegmentationType is LABELMAP (Sup 243)

241111:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Ion Range Surface Compensator (CP 2401), Isodose levels for RT Dose visualization (CP 2402)

241105:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Ocular Region Imaged Module in Ophthalmic Tomography (CP 2346)

241003:	libsrc/standard/condn.tpl,module/base.tpl: Restore not requiring SegmentationMacro if TILED_FULL (CP 2331) lost when adding Label Map Segmentation

240915:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Add SOP Class and data elements for Height Map Segmentation and Revised Ophthalmic OCT En Face Image (Sup 240)

240914:	libsrc/standard/condn.tpl,sopcl.tpl,iodcomp,module,strval/base.tpl: Add Label Map Segmentation, and remove related experimental profile (Sup 243)

240913:	libsrc/standard/elmdict/dicom3.tpl: Add data element for Clarify Description related to Range Shifter Settings reusing DICONDE Material Density (CP 2373 FT2)

240912:	libsrc/standard/elmdict/dicom3.tpl: Add data element for Differentiate acquisition techniques for RT 3D CBCT imaging (CP 2220)

240822: libsrc/standard/module/base.tpl:	Correctly Add Ethnic Group Code Sequence (CP 2356)

240728: libsrc/standard/condn.tpl,module/base.tpl: Report error when PositionReferenceIndicator not SLIDE_CORNER for microscopy (000605)

240712:	appsrc/dcfile/dcacqmap.script,dcacqmap.man: Fix invalid character range in egrep for position files (000604), typo in documentation

240623:	libsrc/standard/elmdict/diconde.tpl: Add DICONDE data elements for thermography imaging

240612:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Table Top Position Alignment UID to Table Top positions (CP 2376)

240612:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Improve RT equipment mapping for imaging (CP 2374)

240611:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl:	Add Ethnic Group Code Sequence (CP 2356)

240610:	libsrc/standard/elmdict/dicom3.tpl:	Add data element for updated RT Patient Setup Module (CP 2344)

240515:	libsrc/standard/condn.tpl,module/base.tpl: Check ICCProfile and ColorSpace not present in top level dataset Image Pixel Module if Optical Path Module present (000603)

240506:	libsrc/standard/module/vl.tpl: Check AnnotationGroupNumber is not zero (000602)

240324:	libsrc/standard/condn.tpl,iodcomp/base.tpl: Detect FrameOfReference Module if FoR UID present even if position and orientation not, since MBPO in Secondary Capture IOD (CP 2330)

240402:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check AnnotationGroupUID is not duplicated (000601)

240329:	libsrc/standard/elmdict/other.tpl: 3DHISTECH private data elements from conformance statement

240329:	appsrc/dcfile/dcentvfy.cc,libsrc/src/locale/mesgtext.tpl: Check all ReferencedSOPInstanceUID instances are in set of supplied files (000600)

240328:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check sanity of Microscopy Bulk Simple Annotation indices and coordinates (000599)

240328:	libsrc/include/dctool/attr.h,attrtype.h,attrtypo.h,src/dctool/attrmxrd.cc,attrtypo.cc: Add support of in-memory (small) OL VR data elements to get values to check (000599)

240325:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl: Clinical Trial Identifiers need Issuers and may be multiple (CP2335)

240324:	libsrc/standard/condn.tpl,iodcomp,module/base.tpl: Clarify Segment scan order for TILED_FULL (CP 2331) and remove CP2331 profile option to test TILED_FULL segmentations

240324:	libsrc/standard/condn.tpl,iodcomp/base.tpl: Add Spatial Information to Secondary Capture IOD (CP 2330)

240324:	libsrc/standard/elmdict/dicom3.tpl,module/vl.tpl: Add information about Vertex Distance to refraction correction parameters (CP 2324)

240324:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl: Add Synthetic Data Flag (CP 2320)

240316:	appsrc/dcfile/dcentvfy.cc: Check Frame of Reference IE attribute values same for same Series (000597)

240312:	libsrc/standard/elmdict/philips.tpl: Add more observed private data elements

240224:	appsrc/dcfile/dciodvfy.cc,libsrc/include/dctool/attr.h,attrothr.h,transyn.h,src/dctool/transyn.cc,transyne.h,locale/mesgtext.tpl,standard/transyn.tpl,support/transyn.awk: Add check that Basic Offset Table is valid if present, including length, point to Item Tag and begin with a JPEG/J2K SOI marker (000531) and point to even offsets (000526)

240222:	appsrc/acrnema/andump.cc,libsrc/include/dctool/attr.h,attrtype.h,attrtypn.h,attrtypo.h,dcstream.h,vr.h,generic/bnstream.h,src/dctool/attr.cc,attrmxrd.cc,attrnew.cc,attrtypo.cc: Add 64 bit binary VRs (CP 1819)

240215:	libsrc/standard/condn.tpl,module/vl.tpl: Correct spurious report of invalid CommonZCoordinateValue for nnotationCoordinateType not 3D when not actually present (000596)

240214:	appsrc/dcfile/dcentvfy.cc: Check Equipment IE attribute values same for same Series (000595)

240131:	libsrc/standard/condn.tpl,iodcomp,module,strval/base.tpl: Add profile option to test LABELMAP segmentations (Sup243)

240118:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check that SCOORD3D POLYGON are closed (first and last points same) (000592)

240114:	libsrc/standard/elmdict/dicom3.tpl: Provide additional ROI parameters (CP 2296)

240112:	libsrc/standard/elmdict/dicom3.tpl: Add new data elements for DICOMweb Storage Commitment Service (Sup 234)

231125:	config/Configure: Build two architectures when clang >=15 but not with ld flag "-force_cpusubtype_ALL" that is no longer supported

231115:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Add Confocal Microscopy data elements and SOP Class UIDs (Sup 236)


appsrc/acrnema/andump.cc
static const char *CopyrightIdentifier(void) { return "@(#)andump.cc Copyright (c) 1993-2024, David A. Clunie DBA PixelMed Publishing. All rights reserved."; }
#if USESTANDARDHEADERSWITHOUTEXTENSION == 1
#include <iostream>
#include <iomanip>
#include <cctype>
#else
#include <iostream.h>
#include <iomanip.h>
#include <ctype.h>
#endif

#if EMITUSINGSTDNAMESPACE == 1
using namespace std;
#endif

#ifdef POWINTEGEREXPONENTTYPE
#define	powi(m,e)	pow(m,(POWINTEGEREXPONENTTYPE)(e))
#else
#define	powi(m,e)	pow(m,(long)(e))
#endif

#ifdef __SC__
#define infinity()    HUGE_VAL
#define quiet_nan(x)  0
#endif /* __SC__ */

#ifdef __MWERKS__
#define infinity()    __inf()
#define quiet_nan(x)  0
#endif /* __MWERKS__ */

#ifdef USEDUMBINFINITYANDNAN
#if USEDUMBINFINITYANDNAN
#define infinity()    HUGE_VAL
#define quiet_nan(x)  0
#endif
#endif

#include "attrtag.h"
#include "dcstream.h"
#include "dcopt.h"
#include "elmconst.h"
#include "elmdict.h"
#include "mesgtext.h"
#include "vr.h"

#ifndef MAXVLTODUMP
#define MAXVLTODUMP 32768
#endif

static TextOutputStream& 
writebase(TextOutputStream& stream,ElementDictionary& dict,const Tag tag,const char *vr,Uint32 vl)
{
	tag.write(stream,&dict);
	stream << "\t VR=<" << (vr ? vr : "") << ">  " << flush;
	stream << " VL=<";
	writeZeroPaddedHexNumber(stream,vl,4);
	stream << ">  " << flush;
	return stream;
}

static TextOutputStream&
doStringValues(DicomInputStream& in,TextOutputStream& log,Uint32 vl,const char * &rstr)
{
	static char buffer[MAXVLTODUMP];
	Assert(vl < MAXVLTODUMP);
	if (vl) in.read(buffer,(int)vl);
	buffer[vl]=0;
	if (!in.fail()) {
		log << "<";
		char *ptr=buffer;
		while (vl--) {
			if (*ptr)
				log << *ptr;
			else if (vl) {	// Allow single trailing null eg. for UI string
				log << endl << WMsgDC(UnexpectedNullInString) << endl;
				break;
			}
			++ptr;
		}
		log << "> ";
		// return value has no leading or trailing space
		ptr=buffer+strlen(buffer);
		while (ptr >= buffer && (*ptr == 0 || isspace(*ptr))) *ptr--=0;
		rstr=buffer;
		while (*rstr && isspace(*rstr)) ++rstr;
	}
	else {
		cerr << EMsgDC(ReadFailed) << endl;
		rstr=0;
	}
	return log;
}

static TextOutputStream&
doTagValues(DicomInputStream& in,TextOutputStream& log,Uint32 vl)
{
	if (vl%4) {
		cerr << EMsgDC(BadValueLength) << endl;
		in.seekg(vl,ios::cur);
		if (in.fail()) {
			cerr << EMsgDC(SeekFailed) << endl;
		}
	}
	else {
		log << "{";
		while (vl) {
			Uint16 group;
			Uint16 element;
			in >> group;
			in >> element;
			log << "(";
			writeZeroPaddedHexNumber(log,group,4);
			log << ",";
			writeZeroPaddedHexNumber(log,element,4);
			log << ")";
			vl-=4;
		}
		log << "} ";
		if (in.fail()) {
			cerr << EMsgDC(ReadFailed) << endl;
		}
	}
	return log;
}

static TextOutputStream&
doBinaryValues(DicomInputStream& in,TextOutputStream& log,Uint64 vl,size_t size,Uint64& rval)
{
	rval=0;
	if (vl < size) size=vl;	// handles special case of 1 US where UL wanted
	if (vl%size) {
		cerr << EMsgDC(BadValueLength) << endl;
		in.seekg(vl,ios::cur);
		if (in.fail()) {
			cerr << EMsgDC(SeekFailed) << endl;
		}
	}
	else {
		log << "[";
		while (vl) {
			Uint64 value;
			switch(size) {
				case 1: { Uint8  v; in >> v; value=v; } break;
				case 2: { Uint16 v; in >> v; value=v; } break;
				case 4: { Uint32 v; in >> v; value=v; } break;
				case 8: { Uint64 v; in >> v; value=v; } break;
				default: Assert(0); break;
			}
			rval=value;
			writeZeroPaddedHexNumber(log,value,size*2);
			vl-=size;
			if (vl) log << ",";
		}
		log << "] ";
		if (in.fail()) {
			cerr << EMsgDC(ReadFailed) << endl;
		}
	}
	return log;
}

static TextOutputStream&
doFloatValues(DicomInputStream& in,TextOutputStream& log,Uint32 vl,size_t size)
{
	if (vl%size) {
		Uint64 ival;
		doBinaryValues(in,log,vl,1,ival);
		cerr << EMsgDC(BadValueLength) << endl;
		//in.seekg(vl,ios::cur);
		//if (in.fail()) {
		//	cerr << EMsgDC(SeekFailed) << endl;
		//}
	}
	else {
		log << "{";
		while (vl) {
			if (size == 4) {
				Uint32 binary; in >> binary;

				double value;
				Int16 sign	=(Int16)((binary&0x80000000)>>31);
				Int16 exponent	=(Int16)((binary&0x7f800000)>>23);
				Uint32 mantissa	= binary&0x007fffff;

				if (exponent == 0) {
					if (mantissa == 0)
						value=0;
					else {
						value=((double)mantissa/
							(1<<23))*powi(2.0,-126);
						value = (sign == 0) ? value : -value;
					}
				}
				else if (exponent == 255) {
					if (mantissa)
						value=quiet_nan(0);
					else
						value=infinity();
				}
				else {
					value=(1.0+(double)mantissa/
						(1<<23))*powi(2.0,exponent-127);
					value = (sign == 0) ? value : -value;
				}
				log << dec << value;
			}
			else if (size == 8) {
				Uint64 binary; in >> binary;

				double value;
						     
				Int16 sign	=(Int16)((binary&0x8000000000000000)>>63);
				Int16 exponent	=(Int16)((binary&0x7ff0000000000000)>>52);
				Uint64 mantissa = binary&0x000fffffffffffff;

				if (exponent == 0) {
					if (mantissa == 0)
						value=0;
					else {
						value=(mantissa/powi(2.0,52))
							*powi(2.0,-1022);
						value = (sign == 0) ? value : -value;
					}
				}
				else if (exponent == 255) {
					if (mantissa)
						value=quiet_nan(0);
					else
						value=infinity();
				}
				else {
					value=(1.0+mantissa/powi(2.0,52))
						*powi(2.0,exponent-1023);
					value = (sign == 0) ? value : -value;
				}
				log << dec << value;
			}
			else {
				Assert(0);
			}

			vl-=size;
			if (vl) log << ",";
		}
		log << "} ";
		if (in.fail()) {
			cerr << EMsgDC(ReadFailed) << endl;
		}
	}
	return log;
}

static bool
readAll(DicomInputStream& stream,TextOutputStream& log,ElementDictionary& dict,bool showoffset,ios::fmtflags showoffsetbase,bool ancreate,bool extractCSA,bool skipiftoolong)
{
	Uint32 n=stream.haveMetaHeader() ? 132 : 0;
	Uint32 length=0xffffffff;
	bool lastwasencapOXorItem=false;
	bool donewithmetaheader=false;
	Tag *siemensCSAHeaderInfoTag = NULL;
	while (stream.peek() != istream::traits_type::eof() && (length == 0xffffffff || n < length)) {
		if (showoffset) {
			log << "@";
			writeZeroPaddedNumber(log,n,showoffsetbase,8);
#ifdef CHECKFORSGINATIVECCSEEKBUG
			log << "(";
			writeZeroPaddedHexNumber(log,stream.rdbuf()->PubSeekOff(0,ios::cur),8);
			log << ")";
#endif
			log << ": ";
		}
		Tag tag;
		stream >> tag;
		if (stream.fail()) {
			cerr << EMsgDC(TagReadFailed) << endl;
			return false;
		}
		n+=4;

		if (!donewithmetaheader && !tag.isMetaheaderGroup()) {
			{
				// backup and read again in case endian changed ...
#ifdef CHECKFORSGINATIVECCSEEKBUG
log << endl << "Before endian check backup @";
writeZeroPaddedHexNumber(log,n,8);
log << "(";
writeZeroPaddedHexNumber(log,stream.rdbuf()->PubSeekOff(0,ios::cur),8);
log << ")"
    << ": ";
#endif
				stream.seekg(-4l,ios::cur);
				if (stream.fail()) {
					cerr << EMsgDC(TagReadFailed) << " - " << EMsgDC(SeekFailed) << endl;
					return false;
				}
#ifdef CHECKFORSGINATIVECCSEEKBUG
log << endl << "Before endian check backup @";
writeZeroPaddedHexNumber(log,n,8);
log << "(";
writeZeroPaddedHexNumber(log,stream.rdbuf()->PubSeekOff(0,ios::cur),8);
log << ")"
    << ": " << endl;
#endif
				if (!stream.getTransferSyntaxToReadDataSet()) {
					cerr << EMsgDC(NoTransferSyntaxInMetaHeader) << endl;
					bool setswapped32big;
					stream.guessTransferSyntaxToReadDataSet(setswapped32big);		// always guesses something
				}
				stream.readingDataSet();
				donewithmetaheader=true;
				stream >> tag;
				if (stream.fail()) {
					cerr << EMsgDC(TagReadFailed) << endl;
					return false;
				}
			}
		}

		char vre[3];
		const char *vr;
		Uint32 vl;

		if (tag == TagFromName(ItemDelimitationItem)
		 || tag == TagFromName(SequenceDelimitationItem)
		 || tag == TagFromName(Item)
		 || tag == Tag(0xfeff,0x00e0)				// GE bug (endianness of Item tag in encapsulated pixel data)
		 || tag == Tag(0xfeff,0xdde0)				// GE bug (endianness of Sequence Delimitation Item tag in encapsulated pixel data)
		) {
			vr=0;
			stream >> vl;
			if (tag == Tag(0xfeff,0x00e0)) {
				vl = ((vl>>24)&0xff) + (((vl>>16)&0xff)<<8) + (((vl>>8)&0xff)<<16) +((vl&0xff)<<24);	// GE bug - swap VL endianness as well
			}
			n+=4;
		}
		else  {
			if (stream.getTransferSyntaxInUse()->isExplicitVR()) {
				stream.read(vre,2);
				if (stream.fail()) {
					cerr << EMsgDC(VRReadFailed) << endl;
					return false;
				}
				if (vre[0] == '-' && vre[1] == '-') {			// illegal proprietary equivalent of UN VR used by Australian Central Data Networks (CDN) PACS
					vr="UN";
					vl=stream.read16();
					n+=2;
				}
				else if (vre[0] == 0 && vre[1] == 0) {			// illegal two bytes of zero VR encountered in Sante de-identifier output (followed by two byte length)
					vr="UN";
					vl=stream.read16();
					n+=2;
				}
				else if (vre[0] < 'A' || vre[1] < 'A') {		// DicomWorks bug ... occasional implicit VR attribute even though explicit transfer syntax
					//cerr << EMsgDC(TagReadFailed) << " - " << MMsgDC(ImplicitVREncodingWhenExplicitRequired) << endl;
					vr="UN";
					stream.seekg(-2l,ios::cur);
					if (stream.fail()) {
						cerr << EMsgDC(TagReadFailed) << " - " << EMsgDC(SeekFailed) << endl;
						return false;
					}
					stream >> vl;				// read VL as if implicit VR
					n+=4;
				}
				else {
					n+=2;
					vre[2]=0;
					vr=vre;
					if (isUniversalResourceVR(vr) || isUnlimitedCharactersVR(vr) || isUnlimitedTextVR(vr) || isUnknownVR(vr)) {
						// should be long form, but check for bad implementations that use short form ...
						vl=stream.read16();
						n+=2;
						if (vl == 0) {		// if reserved bytes not zero, assume used for short form of length (risky) :(
							stream >> vl;
							n+=4;
						}
						else {
							cerr << EMsgDC(IncorrectShortVLForUROrUCOrUTOrUN) << endl;
						}
					}
					else if (isLongValueLengthInExplicitValueRepresentation(vr)) {	// some of these already handled above
						(void)stream.read16();	// "Reserved"
						stream >> vl;
						n+=6;
					}
					else {
						vl=stream.read16();
						n+=2;
					}
				}
			}
			else {
				vr=(tag.isPrivateGroup() && tag.isPrivateOwner())
					? "LO"	// PS3.5-7.8.1
					: dict.getValueRepresentation(tag);
				stream >> vl;
				n+=4;
			}
		}
		if (stream.fail()) {
			cerr << EMsgDC(VLReadFailed) << endl;
			return false;
		}

		if (ancreate) {
			if (tag == TagFromName(ItemDelimitationItem)) {
				log << "%enditem";
			}
			else if (tag == TagFromName(SequenceDelimitationItem)) {
				log << "%endseq";
			}
			else if (tag == TagFromName(Item)) {
				log << "%item";
			}
			else {
				writebase(log,dict,tag,vr,vl);
			}
		}
		else {
			writebase(log,dict,tag,vr,vl);
		}
		
		if (vl == 0xffffffff		// SQ or OX both will contain items
		 || vl == 0			// includes ItemDelimitationItem,SequenceDelimitationItem
		 || isSequenceVR(vr)		// SQ with defined length
		 || (tag == TagFromName(Item) && !lastwasencapOXorItem)	// Item with element contents
		) {
			if (!ancreate
			 || (tag != TagFromName(ItemDelimitationItem)
			  && tag != TagFromName(SequenceDelimitationItem)
			  && tag != TagFromName(Item))) {
				log << "\t[]";
			}
		}
		else {
			if (extractCSA && siemensCSAHeaderInfoTag && *siemensCSAHeaderInfoTag == tag) {
				// do nothing - treat it as containing DICOM tags and loop around - works for EVRLE only
				// should also check to stop at (0xffff,0xffff)
			}
			else if (vl < MAXVLTODUMP) {
				if (isStringVR(vr)) {
					const char *sval;
					doStringValues(stream,log,vl,sval);
					if (!stream.fail() && vl && sval && *sval) {
						if (tag == TagFromName(TransferSyntaxUID)) {
							if (!donewithmetaheader) {
								stream.setTransferSyntaxToReadDataSet(new TransferSyntax(sval));
							}
						}
						else if (tag.isPrivateGroup() && tag.isPrivateOwner()) {
							dict.addOwner(tag,sval);
							if (tag.getGroup() == 0x0029 && strcmp(sval,"SIEMENS CSA HEADER") == 0) {
								Uint16 siemensCSAHeaderInfoElement=((tag.getElement() << 8) & 0xff00)+0x0010;
								siemensCSAHeaderInfoTag = new Tag(0x0029,siemensCSAHeaderInfoElement);
							}
						}
					}
				}
				else if (isAttributeTagVR(vr))
					doTagValues(stream,log,vl);
				else if (isNumericVR(vr)) {
					Uint64 ival;
					size_t size=sizeofNumericVR(vr);
					doBinaryValues(stream,log,vl,size,ival);
					if (!stream.fail() && vl == size) {
						if (tag == TagFromName(LengthToEnd)
						 && length == 0xffffffff) {
							// Retired, but if you've got it use it
							//length=n+ival+vl;
						}
					}
				}
				else if (isFloatVR(vr)) {
					size_t size=sizeofFloatVR(vr);
					doFloatValues(stream,log,vl,size);
				}
				else {
					//cerr << endl << WMsgDC(UnrecognizedVR) << endl;
					Uint64 ival;
					doBinaryValues(stream,log,vl,1,ival);
				}
			}
			else {
				Assert(!ancreate || skipiftoolong);
				if (ancreate) {
					log << "\t[...]";
				}
				else {
					log << "\t[]\t# skipping ...";
				}
				stream.seekg(vl,ios::cur);
				if (stream.fail()) {
					cerr << EMsgDC(SeekFailed) << endl;
					return false;
				}
			}
			n+=vl;
		}
		lastwasencapOXorItem=
			(isOtherByteOrWordOrUnspecifiedVR(vr)
			 || tag == TagFromName(Item)
			 || tag == TagFromName(ItemDelimitationItem)
			) && stream.getTransferSyntaxInUse()->isEncapsulated();

		log << endl;
	}
	return true;
}

int
main(int argc, char *argv[])
{
	GetNamedOptions    options(argc,argv);
	DicomInputOptions input_options(options);

	bool showfilename=options.get("filename");
	bool ancreate=options.get("ancreate");
	bool extractCSA=options.get("extractCSA");
	bool skipiftoolong=options.get("skipiftoolong");
	bool showoffset=false;
	ios::fmtflags showoffsetbase;
	if (options["showoffset"]||options["showoffset-hex"])
		{ showoffset=true; showoffsetbase=ios::hex; }
	else if (options["showoffset-octal"]||options["showoffset-oct"])
		{ showoffset=true; showoffsetbase=ios::oct; }
	else if (options["showoffset-decimal"]||options["showoffset-dec"])
		{ showoffset=true; showoffsetbase=ios::dec; }

	input_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,input_options.filename,cin);

	cerr << options.errors();
	cerr << input_options.errors();
	cerr << input_opener.errors();

	if (!input_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< input_options.usage()
			<< " [-ancreate]"
			<< " [-extractCSA]"
			<< " [-skipiftoolong]"
			<< " [-showoffset|-showoffset-hex"
			<< "|-showoffset-octal|-showoffset-oct"
			<< "|-showoffset-decimal|-showoffset-dec]"
			<< " [-filename]"
			<< " [" << MMsgDC(InputFile) << "]"
			<< " <" << MMsgDC(InputFile)
			<< endl;
		exit(1);
	}

	DicomInputStream din(*(istream *)input_opener,
		input_options.transfersyntaxuid,
		input_options.usemetaheader);
	TextOutputStream log(cerr);

	if (showfilename) {
		const char *filenameused = input_opener.getFilename();
		log << "Filename: \"" << (filenameused && strlen(filenameused) > 0 ? filenameused : "-") << "\"" << endl;
	}

	ElementDictionary dictionary;

	readAll(din,log,dictionary,showoffset,showoffsetbase,ancreate,extractCSA,skipiftoolong);
}
